src/main/java/hudson/remoting/AtmostOneThreadExecutor.java

```diff
@@ -71,10 +71,11 @@ public boolean isTerminated() {
 
     public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
         synchronized (q) {
-            long start = System.currentTimeMillis();
-            long end = start+unit.toMillis(timeout);
-            while (isAlive() && System.currentTimeMillis()<end) {
-                q.wait(end-System.currentTimeMillis());
+            long now = System.nanoTime();
+            long end = now + unit.toNanos(timeout);
+            while (isAlive() && (end - now > 0L)) {
+                q.wait(TimeUnit.NANOSECONDS.toMillis(end - now));
```

Comment:
Could use the two-arg overload of `Object.wait` to specify a precise nanosecond time, though at best you could save a millisecond of waiting time that way, so maybe not worth the effort.